Create minimal main display item palette (#1176) (#1193)
AlexsLemonade/OpenPBTA-analysis: figures/mapping-histology-labels.Rmd

---
title: "Create a minimal palette for displaying multiple disease labels"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen, Krutika Gaonkar, and Jaclyn Taroni
date: 2021
---

## Purpose

There are multiple "disease labels" in the `pbta-histologies.tsv` file, including (from most broad to most narrow) `broad_histology`, `cancer_group`, and `harmonized_diagnosis`.
For context, it is helpful to note that an individual `cancer_group` will be nested under a single `broad_histology` and that `cancer_group` is a shorter form of `harmonized_diagnosis` with the following edits:

> - Removed Other, Benign tumor and Dysplasia/Gliosis, Dysplasia/Gliosis-Glial-neuronal tumor NOS removed from `cancer_group`

Reviewer: Redundant wording
```suggestion
- Other, Benign tumor and Dysplasia/Gliosis, Dysplasia/Gliosis-Glial-neuronal tumor NOS removed from `cancer_group`
```

Check out the discussion on [#1174](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/1174)!

### `broad_histology`

```{r}
broad_histology_df <- data.frame(
  broad_histology = c("Benign tumor",
                      "Diffuse astrocytic and oligodendroglial tumor",
                      "Embryonal tumor",
                      "Ependymal tumor",
                      "Germ cell tumor",
                      "Low-grade astrocytic tumor",
                      "Meningioma",
                      "Mesenchymal non-meningothelial tumor",
                      "Neuronal and mixed neuronal-glial tumor",
                      "Tumor of cranial and paraspinal nerves",
                      "Tumors of sellar region"), 
  broad_histology_hex = c("#590024",
                          "#ff80e5",
                          "#220040",
                          "#2200ff",
                          "#0074d9",
                          "#8f8fbf",
                          "#2db398",
                          "#7fbf00",
                          "#685815",
                          "#ffaa00",
                          "#b2502d"),
  stringsAsFactors = FALSE

> )

Reviewer: Filling in my suggestion. 
```suggestion
)
# value for "other" histologies
broad_histology_other_hex = "#808080"
```

                   "High-grade glioma astrocytoma",
                   "Atypical Teratoid Rhabdoid Tumor",
                   "CNS Embryonal tumor",
                   "Medulloblastoma",
                   "Ependymoma",
                   "Teratoma",
                   "Ganglioglioma",
                   "Low-grade glioma astrocytoma",
                   "Meningioma",
                   "Ewing sarcoma",
                   "Dysembryoplastic neuroepithelial tumor",
                   "Neurofibroma Plexiform",
                   "Schwannoma",
                   "Craniopharyngioma"),
  cancer_group_hex = c("#4d2635",
                       "#bf0099",
                       "#ff40d9",
                       "#ffccf5",
                       "#4d0d85",
                       "#b08ccf",
                       "#a340ff",
                       "#2200ff",
                       "#058aff",
                       "#8c8cff",
                       "#000080",
                       "#2db398",
                       "#9fbf60",
                       "#614e01",
                       "#e6ac39",
                       "#ab7200",
                       "#b33000"),
  stringsAsFactors = FALSE

> )

Reviewer: Filling in my suggestion 
```suggestion
)
# Value for "other" groups
cancer_group_other_hex == "#b5b5b5"
```

```

And again, we'll create a legend with `legend()`

```{r}
plot(NULL, xaxt = "n", yaxt = "n", bty = "n", ylab = "", xlab = "", 
     xlim = 0:1, ylim = 0:1)
legend("topleft", 
       legend = cancer_group_df$cancer_group,
       col = cancer_group_df$cancer_group_hex,
       pch = 15, pt.cex = 1.5, cex = 0.75, bty = "n")
mtext("Cancer Group", at = 0.0625, cex = 1)
```

### Output

> We can create a data frame that contains both palettes with a series of left joins, where we will then fill the NA values with a single (gray) hex code per column (`#808080` for `broad_histology`, `#d3d3d3` for `cancer_group`.)

Reviewer: Got fancy with inserting the colors by variable name here. (Noting that even if you don't do this, the text should be updated to the correct `cancer_group` color.)
```suggestion
We can create a data frame that contains both palettes with a series of left joins, where we will then fill the NA values with a single (gray) hex code per column (`r broad_histology_other_hex` for `broad_histology`, `r cancer_group_other_hex` for `cancer_group`.)
```

```{r}
palette_df <- histologies_df %>%
  # Exclude normal samples
  filter(sample_type == "Tumor") %>%
  # Filter to unique broad histology--cancer group pairs
  select(broad_histology, 
         cancer_group) %>% 
  distinct() %>%
  # Add broad histology palette
  left_join(broad_histology_df, by = "broad_histology") %>%
  # Add cancer group palette
  left_join(cancer_group_df, by = "cancer_group") %>%
  # Fill all other values with gray colors

>   replace_na(list(broad_histology_hex = "#808080", 
>                   cancer_group_hex = "#b5b5b5")) %>%

Reviewer: Because we use these twice, maybe we should pull them out to variables (defined earlier, with the data frames, presumably).
```suggestion
  replace_na(list(broad_histology_hex = broad_histology_other_hex,
                  cancer_group_hex = cancer_group_other_hex)) %>%
```

  # The exception being - if cancer_group == NA, so should cancer_group_hex!
  mutate(cancer_group_hex = if_else(is.na(cancer_group), 
                                    NA_character_, 
                                    cancer_group_hex)) %>%
  # Sort by broad_histology for easy browsing
  arrange(broad_histology)
```

And now let's take a look!

```{r}
palette_df
```

### Add display names for convenience

When multiple values are using the same color, it can be helpful to have a separate value for the legend, e.g., for all `#808080` broad histologies, we may want to display `Other`. 
We'll add a couple columns for legend-making convenience.

```{r}
palette_df <- palette_df %>%

>   mutate(broad_histology_display = if_else(broad_histology_hex == "#808080",
>                                            "Other",
>                                            broad_histology),
>          cancer_group_display = if_else(cancer_group_hex == "#b5b5b5",

Reviewer: Using my earlier suggested variables
```suggestion
  mutate(broad_histology_display = if_else(broad_histology_hex == broad_histology_other_hex,
                                           "Other",
                                           broad_histology),
         cancer_group_display = if_else(cancer_group_hex == cancer_group_other_hex,
```

  ))
```

For cancer groups that do not get their own hex code for display (i.e., due to small sample sizes), we'll use a selection of grey colors as a palette and rely heavily on the ordering of the OncoPrint legend.
Unfortunately there are over 20 "Other CNS" cancer groups that meet this criterion, so it is not feasible to have a color for each of them and they will not be included in the "Other CNS" OncoPrint.

```{r}
greys_df <- palette_df %>%
  filter(cancer_group_display == "Other",
         !is.na(oncoprint_group),
         oncoprint_group != "Other CNS")

# Sample the greys sequential palette from color brewer
set.seed(2021)
greys_df <- greys_df %>%
  mutate(oncoprint_hex = sample(brewer.pal(nrow(greys_df), "Greys")))
```

Add `oncoprint_hex` & `oncoprint_include`, the latter will be `FALSE` when the former is `NA`.

Add `oncoprint_hex` for all other cancer groups and `oncoprint_include`.
The latter will be FALSE when the former is NA.

```{r}
palette_df <- palette_df %>%
  left_join(greys_df) %>%
  # When there's an oncoprint group and a specific cancer group display color,
  # use the cancer group display color as the oncoprint color
  mutate(oncoprint_hex = if_else(

>     cancer_group_hex != "#b5b5b5" & !is.na(oncoprint_group),

Reviewer: variable substitution
```suggestion
    cancer_group_hex != cancer_group_other_hex & !is.na(oncoprint_group),
```
